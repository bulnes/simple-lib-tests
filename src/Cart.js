import Dinero from 'dinero.js';
import { calculateDiscount } from './discount.utils';

const Money = Dinero;

Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

export class Cart {
  items = [];

  add(item) {
    const productToFind = item.product;
    const itemIndex = this.items.findIndex(item => {
      return item.product.title === productToFind.title;
    });

    if (itemIndex === -1) {
      this.items.push(item);
    }
  }

  remove(product) {
    const itemIndex = this.items.findIndex(item => {
      return item.product.title === product.title;
    });

    if (itemIndex !== -1) {
      this.items.splice(itemIndex, 1);
    } else {
      throw new Error('Product not found');
    }
  }

  getTotal() {
    return this.items.reduce((total, { product, condition, quantity }) => {
      const amount = Money({ amount: product.price * quantity });

      const discount = condition
        ? calculateDiscount(amount, quantity, condition)
        : Money({ amount: 0 });

      return total.add(amount).subtract(discount);
    }, Money({ amount: 0 }));
  }

  summary() {
    const total = this.getTotal().getAmount();
    const formatted = this.getTotal().toFormat('$0,0.00');
    const items = this.items;

    return { total, formatted, items };
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];

    return { total, items };
  }
}
