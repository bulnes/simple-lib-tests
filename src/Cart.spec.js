import { Cart } from './Cart';

describe('Cart', () => {
  let cart;

  let product = {
    title: 'Adidas shoes - men',
    price: 35388,
  };

  let product2 = {
    title: 'Adidas shoes - women',
    price: 40000,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 to empty product list', () => {
      expect(cart.getTotal().getAmount()).toBe(0);
    });

    it('should return the sum of all products and quantities', () => {
      const item = {
        product,
        quantity: 2,
      };

      cart.add(item);

      expect(cart.getTotal().getAmount()).toBe(70776);
    });

    it('should have one product at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toBe(70776);
    });

    it('should ensure that total get updated when a product is removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toBe(110776);

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toBe(40000);
    });
  });

  describe('remove()', () => {
    it('should throw an error when trying to remove a product that does not exist', () => {
      expect(() => cart.remove(product)).toThrow();
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should get an empty list of products after a checkout', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.checkout();

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe('summary()', () => {
    it('should return an object with the total and the items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.summary()).toMatchSnapshot();
    });

    it('should return an object with the formatted total', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.summary().formatted).toEqual('R$1,107.76');
    });
  });

  describe('special conditions', () => {
    it('should apply percentage discount quantity above minimum amount of products', () => {
      const condition = {
        percentage: 30, // 30%
        minimum: 2,
      };

      cart.add({
        product,
        quantity: 3,
        condition,
      });

      expect(cart.getTotal().getAmount()).toBe(74315);
    });

    it('should not apply percentage discount quantity above minimum amount of products', () => {
      const condition = {
        percentage: 30, // 30%
        minimum: 2,
      };

      cart.add({
        product,
        quantity: 2,
        condition,
      });

      expect(cart.getTotal().getAmount()).toBe(70776);
    });

    it('should apply 50% percentage discount quantity above minimum amount of products', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        quantity: 4,
        condition,
      });

      expect(cart.getTotal().getAmount()).toBe(70776);
    });

    it('should not apply 50% percentage discount quantity above minimum amount of products', () => {
      const condition = {
        quantity: 3,
      };

      cart.add({
        product,
        quantity: 3,
        condition,
      });

      expect(cart.getTotal().getAmount()).toBe(106164);
    });

    it('should apply 40% percentage discount quantity is odd', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        quantity: 5,
        condition,
      });

      expect(cart.getTotal().getAmount()).toBe(106164);
    });

    it('should apply higher percentage of discount. First case.', () => {
      const condition1 = {
        percentage: 30, // 30%
        minimum: 2,
      };

      const condition2 = {
        quantity: 2, // 40%
      };

      cart.add({
        product,
        quantity: 5,
        condition: [condition1, condition2],
      });

      expect(cart.getTotal().getAmount()).toBe(106164);
    });

    it('should apply higher percentage of discount. Second case.', () => {
      const condition1 = {
        percentage: 50, // 50%
        minimum: 2,
      };

      const condition2 = {
        quantity: 2, // 40%
      };

      cart.add({
        product,
        quantity: 5,
        condition: [condition1, condition2],
      });

      expect(cart.getTotal().getAmount()).toBe(88470);
    });

    it('should apply higher percentage of discount. Thrid case.', () => {
      const condition1 = {};

      const condition2 = {};

      cart.add({
        product,
        quantity: 5,
        condition: [condition1, condition2],
      });

      expect(cart.getTotal().getAmount()).toBe(176940);
    });
  });
});
