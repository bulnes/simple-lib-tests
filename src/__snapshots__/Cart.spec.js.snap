// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas shoes - men",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 40000,
        "title": "Adidas shoes - women",
      },
      "quantity": 1,
    },
  ],
  "total": 110776,
}
`;

exports[`Cart summary() should return an object with the total and the items 1`] = `
{
  "formatted": "R$1,107.76",
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas shoes - men",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 40000,
        "title": "Adidas shoes - women",
      },
      "quantity": 1,
    },
  ],
  "total": 110776,
}
`;
